File: backend/app/schemas/human_resources/promotion_application.py
"""
集团人资基础 - 晋升申请 schemas
"""

from datetime import date, datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

PROMOTION_APPROVAL_STAGES = [
    "department_manager",
    "principal",
    "biz_director",
    "hr_director",
    "chairman",
]

PROMOTION_STAGE_LABELS = {
    "department_manager": "部门主管",
    "principal": "校长",
    "biz_director": "业务条线总监",
    "hr_director": "人资总监",
    "chairman": "董事长",
}

PROMOTION_STATUS_LABELS = {
    "draft": "草稿",
    "pending": "审批中",
    "approved": "已通过",
    "rejected": "已驳回",
}

PROMOTION_FLOW_STATUS_LABELS = {
    "completed": "已通过",
    "current": "审批中",
    "rejected": "已驳回",
    "waiting": "待审批",
}


class PromotionApprovalConfigApproverOut(BaseModel):
    id: int
    approver_user_id: int
    approver_name: str
    approver_department: Optional[str] = None
    approver_position: Optional[str] = None
    approver_campus: Optional[str] = None
    sort_order: int

    model_config = ConfigDict(
        from_attributes=True,
    )


class PromotionApprovalConfigUpsert(BaseModel):
    campus: str = Field(..., min_length=1, max_length=100)
    apply_department: Optional[str] = Field(None, max_length=100)
    apply_position: Optional[str] = Field(None, max_length=100)
    stage: str = Field(..., max_length=50)
    approver_user_ids: List[int] = Field(default_factory=list)
    is_active: bool = True


class PromotionApprovalConfigOut(BaseModel):
    id: int
    campus: str
    apply_department: Optional[str] = None
    apply_position: Optional[str] = None
    stage: str
    stage_label: str
    is_active: bool
    approvers: List[PromotionApprovalConfigApproverOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class PromotionCurrentApproverOut(BaseModel):
    user_id: int
    name: str
    department: Optional[str] = None
    position: Optional[str] = None
    campus: Optional[str] = None


class PromotionApprovalPreviewInput(BaseModel):
    campus: str = Field(..., min_length=1, max_length=100)
    department: str = Field(..., min_length=1, max_length=100)
    position: str = Field(..., min_length=1, max_length=100)


class PromotionApproverCandidateOut(BaseModel):
    stage: str
    stage_label: str
    recommended_user_ids: List[int] = Field(default_factory=list)
    approvers: List[PromotionCurrentApproverOut] = Field(default_factory=list)


class PromotionApplicationActionPayload(BaseModel):
    comment: Optional[str] = Field(None, max_length=2000)
    original_level: Optional[str] = Field(None, max_length=100)
    original_salary: Optional[float] = None
    promoted_level: Optional[str] = Field(None, max_length=100)
    promoted_base_salary: Optional[float] = None
    promoted_performance_salary: Optional[float] = None
    promoted_salary: Optional[float] = None


class PromotionApprovalActionOut(BaseModel):
    id: int
    stage: str
    stage_label: str
    action: str
    approver_user_id: Optional[int] = None
    approver_name: Optional[str] = None
    comment: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class PromotionApprovalFlowStepOut(BaseModel):
    stage: str
    stage_label: str
    status: str
    status_label: str
    approvers: List[PromotionCurrentApproverOut] = Field(default_factory=list)
    action: Optional[str] = None
    action_label: Optional[str] = None
    acted_by_user_id: Optional[int] = None
    acted_by_name: Optional[str] = None
    comment: Optional[str] = None
    acted_at: Optional[datetime] = None


class PromotionNotificationOut(BaseModel):
    id: int
    application_id: int
    application_no: str
    notification_type: str
    title: str
    content: str
    stage: Optional[str] = None
    stage_label: Optional[str] = None
    is_read: bool
    action_by_user_id: Optional[int] = None
    action_by_name: Optional[str] = None
    created_at: datetime
    read_at: Optional[datetime] = None


class PromotionApplicationBase(BaseModel):
    fill_date: date
    campus: str = Field(..., min_length=1, max_length=100)
    name: str = Field(..., min_length=1, max_length=100)
    native_place: Optional[str] = Field(None, max_length=100)
    age: Optional[int] = Field(None, ge=0, le=100)
    entry_date: date
    department: str = Field(..., min_length=1, max_length=100)
    position: str = Field(..., min_length=1, max_length=100)
    work_overview: str = Field(..., min_length=1)
    promotion_reason: str = Field(..., min_length=1)
    confidence_and_expectation: str = Field(..., min_length=1)
    original_level: Optional[str] = Field(None, max_length=100)
    original_salary: Optional[float] = None
    promoted_level: Optional[str] = Field(None, max_length=100)
    promoted_base_salary: Optional[float] = None
    promoted_performance_salary: Optional[float] = None
    promoted_salary: Optional[float] = None
    selected_approver_user_ids: Dict[str, List[int]] = Field(default_factory=dict)


class PromotionApplicationCreate(PromotionApplicationBase):
    pass


class PromotionApplicationUpdate(BaseModel):
    fill_date: Optional[date] = None
    campus: Optional[str] = Field(None, max_length=100)
    name: Optional[str] = Field(None, max_length=100)
    native_place: Optional[str] = Field(None, max_length=100)
    age: Optional[int] = Field(None, ge=0, le=100)
    entry_date: Optional[date] = None
    department: Optional[str] = Field(None, max_length=100)
    position: Optional[str] = Field(None, max_length=100)
    work_overview: Optional[str] = None
    promotion_reason: Optional[str] = None
    confidence_and_expectation: Optional[str] = None
    original_level: Optional[str] = Field(None, max_length=100)
    original_salary: Optional[float] = None
    promoted_level: Optional[str] = Field(None, max_length=100)
    promoted_base_salary: Optional[float] = None
    promoted_performance_salary: Optional[float] = None
    promoted_salary: Optional[float] = None
    selected_approver_user_ids: Optional[Dict[str, List[int]]] = None


class PromotionApplicationOut(PromotionApplicationBase):
    id: int
    application_no: str
    created_by_user_id: Optional[int] = None
    created_by_name: Optional[str] = None
    department_manager_opinion: Optional[str] = None
    department_manager_passed: Optional[bool] = None
    principal_opinion: Optional[str] = None
    principal_passed: Optional[bool] = None
    biz_director_opinion: Optional[str] = None
    biz_director_passed: Optional[bool] = None
    hr_director_opinion: Optional[str] = None
    hr_director_passed: Optional[bool] = None
    chairman_opinion: Optional[str] = None
    chairman_passed: Optional[bool] = None
    is_passed: Optional[bool] = None
    status: str
    status_label: str
    current_stage: Optional[str] = None
    current_stage_label: Optional[str] = None
    rejection_reason: Optional[str] = None
    submitted_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    can_edit: bool = False
    can_delete: bool = False
    can_submit: bool = False
    can_approve: bool = False
    selected_approver_user_ids: Dict[str, List[int]] = Field(default_factory=dict)
    current_approvers: List[PromotionCurrentApproverOut] = Field(default_factory=list)
    approval_flow: List[PromotionApprovalFlowStepOut] = Field(default_factory=list)
    approval_actions: List[PromotionApprovalActionOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )

File: frontend/services/humanresources/promotionApplication.ts
import { api } from '../api'

export type PromotionStatus = 'draft' | 'pending' | 'approved' | 'rejected'
export type PromotionStage =
  | 'department_manager'
  | 'principal'
  | 'biz_director'
  | 'hr_director'
  | 'chairman'

export interface PromotionApproverCandidate {
  userId: number
  name: string
  department?: string | null
  position?: string | null
  campus?: string | null
}

export interface PromotionApprovalPreviewStage {
  stage: PromotionStage
  stageLabel: string
  recommendedUserIds: number[]
  approvers: PromotionApproverCandidate[]
}

interface ApiPromotionCurrentApprover {
  user_id: number
  name: string
  department?: string | null
  position?: string | null
  campus?: string | null
}

interface ApiPromotionApprovalAction {
  id: number
  stage: string
  stage_label: string
  action: string
  approver_user_id?: number | null
  approver_name?: string | null
  comment?: string | null
  created_at: string
}

interface ApiPromotionApprovalFlowStep {
  stage: string
  stage_label: string
  status: string
  status_label: string
  approvers: ApiPromotionCurrentApprover[]
  action?: string | null
  action_label?: string | null
  acted_by_user_id?: number | null
  acted_by_name?: string | null
  comment?: string | null
  acted_at?: string | null
}

interface ApiPromotionApplicationRecord {
  id: number
  application_no: string
  fill_date: string
  campus: string
  name: string
  native_place?: string | null
  age?: number | null
  entry_date: string
  department: string
  position: string
  work_overview: string
  promotion_reason: string
  confidence_and_expectation: string
  original_level?: string | null
  original_salary?: number | null
  promoted_level?: string | null
  promoted_base_salary?: number | null
  promoted_performance_salary?: number | null
  promoted_salary?: number | null
  created_by_user_id?: number | null
  created_by_name?: string | null
  department_manager_opinion?: string | null
  department_manager_passed?: boolean | null
  principal_opinion?: string | null
  principal_passed?: boolean | null
  biz_director_opinion?: string | null
  biz_director_passed?: boolean | null
  hr_director_opinion?: string | null
  hr_director_passed?: boolean | null
  chairman_opinion?: string | null
  chairman_passed?: boolean | null
  is_passed?: boolean | null
  status: PromotionStatus
  status_label: string
  current_stage?: string | null
  current_stage_label?: string | null
  rejection_reason?: string | null
  submitted_at?: string | null
  completed_at?: string | null
  can_edit: boolean
  can_delete: boolean
  can_submit: boolean
  can_approve: boolean
  selected_approver_user_ids: Partial<Record<PromotionStage, number[]>>
  current_approvers: ApiPromotionCurrentApprover[]
  approval_flow: ApiPromotionApprovalFlowStep[]
  approval_actions: ApiPromotionApprovalAction[]
  created_at: string
  updated_at: string
}

export interface PromotionCurrentApprover {
  userId: number
  name: string
  department?: string | null
  position?: string | null
  campus?: string | null
}

export interface PromotionApprovalAction {
  id: number
  stage: string
  stageLabel: string
  action: string
  approverUserId?: number | null
  approverName?: string | null
  comment?: string | null
  createdAt: string
}

export interface PromotionApprovalFlowStep {
  stage: string
  stageLabel: string
  status: string
  statusLabel: string
  approvers: PromotionCurrentApprover[]
  action?: string | null
  actionLabel?: string | null
  actedByUserId?: number | null
  actedByName?: string | null
  comment?: string | null
  actedAt?: string | null
}

export interface PromotionApplicationRecord {
  id: number
  applicationNo: string
  fillDate: string
  campus: string
  name: string
  nativePlace?: string | null
  age?: number | null
  entryDate: string
  department: string
  position: string
  workOverview: string
  promotionReason: string
  confidenceAndExpectation: string
  originalLevel?: string | null
  originalSalary?: number | null
  promotedLevel?: string | null
  promotedBaseSalary?: number | null
  promotedPerformanceSalary?: number | null
  promotedSalary?: number | null
  createdByUserId?: number | null
  createdByName?: string | null
  departmentManagerOpinion?: string | null
  departmentManagerPassed?: boolean | null
  principalOpinion?: string | null
  principalPassed?: boolean | null
  bizDirectorOpinion?: string | null
  bizDirectorPassed?: boolean | null
  hrDirectorOpinion?: string | null
  hrDirectorPassed?: boolean | null
  chairmanOpinion?: string | null
  chairmanPassed?: boolean | null
  isPassed?: boolean | null
  status: PromotionStatus
  statusLabel: string
  currentStage?: string | null
  currentStageLabel?: string | null
  rejectionReason?: string | null
  submittedAt?: string | null
  completedAt?: string | null
  canEdit: boolean
  canDelete: boolean
  canSubmit: boolean
  canApprove: boolean
  selectedApproverUserIds: Partial<Record<PromotionStage, number[]>>
  currentApprovers: PromotionCurrentApprover[]
  approvalFlow: PromotionApprovalFlowStep[]
  approvalActions: PromotionApprovalAction[]
  createdAt: string
  updatedAt: string
}

export interface PromotionApplicationPayload {
  fillDate: string
  campus: string
  name: string
  nativePlace?: string | null
  age?: number | null
  entryDate: string
  department: string
  position: string
  workOverview: string
  promotionReason: string
  confidenceAndExpectation: string
  originalLevel?: string | null
  originalSalary?: number | null
  promotedLevel?: string | null
  promotedBaseSalary?: number | null
  promotedPerformanceSalary?: number | null
  promotedSalary?: number | null
  selectedApproverUserIds?: Partial<Record<PromotionStage, number[]>>
}

export interface PromotionApprovalActionPayload {
  comment?: string
  originalLevel?: string | null
  originalSalary?: number | null
  promotedLevel?: string | null
  promotedBaseSalary?: number | null
  promotedPerformanceSalary?: number | null
  promotedSalary?: number | null
}

interface ApiPromotionApprovalPreviewStage {
  stage: PromotionStage
  stage_label: string
  recommended_user_ids: number[]
  approvers: ApiPromotionCurrentApprover[]
}

interface ApiPromotionApprovalConfigApprover {
  id: number
  approver_user_id: number
  approver_name: string
  approver_department?: string | null
  approver_position?: string | null
  approver_campus?: string | null
  sort_order: number
}

interface ApiPromotionApprovalConfig {
  id: number
  campus: string
  apply_department?: string | null
  apply_position?: string | null
  stage: PromotionStage
  stage_label: string
  is_active: boolean
  approvers: ApiPromotionApprovalConfigApprover[]
  created_at: string
  updated_at: string
}

export interface PromotionApprovalConfig {
  id: number
  campus: string
  applyDepartment?: string | null
  applyPosition?: string | null
  stage: PromotionStage
  stageLabel: string
  isActive: boolean
  approvers: Array<{
    id: number
    approverUserId: number
    approverName: string
    approverDepartment?: string | null
    approverPosition?: string | null
    approverCampus?: string | null
    sortOrder: number
  }>
  createdAt: string
  updatedAt: string
}

export interface PromotionApprovalConfigPayload {
  campus: string
  applyDepartment?: string | null
  applyPosition?: string | null
  stage: PromotionStage
  approverUserIds: number[]
  isActive: boolean
}

const mapApplicationPayload = (payload: PromotionApplicationPayload) => ({
  fill_date: payload.fillDate,
  campus: payload.campus,
  name: payload.name,
  native_place: payload.nativePlace || undefined,
  age: payload.age ?? undefined,
  entry_date: payload.entryDate,
  department: payload.department,
  position: payload.position,
  work_overview: payload.workOverview,
  promotion_reason: payload.promotionReason,
  confidence_and_expectation: payload.confidenceAndExpectation,
  original_level: payload.originalLevel || undefined,
  original_salary: payload.originalSalary ?? undefined,
  promoted_level: payload.promotedLevel || undefined,
  promoted_base_salary: payload.promotedBaseSalary ?? undefined,
  promoted_performance_salary: payload.promotedPerformanceSalary ?? undefined,
  promoted_salary: payload.promotedSalary ?? undefined,
  selected_approver_user_ids: payload.selectedApproverUserIds || {},
})

const mapPartialApplicationPayload = (payload: Partial<PromotionApplicationPayload>) => {
  const result: Record<string, unknown> = {}
  if (payload.fillDate !== undefined) result.fill_date = payload.fillDate
  if (payload.campus !== undefined) result.campus = payload.campus
  if (payload.name !== undefined) result.name = payload.name
  if (payload.nativePlace !== undefined) result.native_place = payload.nativePlace || undefined
  if (payload.age !== undefined) result.age = payload.age ?? undefined
  if (payload.entryDate !== undefined) result.entry_date = payload.entryDate
  if (payload.department !== undefined) result.department = payload.department
  if (payload.position !== undefined) result.position = payload.position
  if (payload.workOverview !== undefined) result.work_overview = payload.workOverview
  if (payload.promotionReason !== undefined) result.promotion_reason = payload.promotionReason
  if (payload.confidenceAndExpectation !== undefined) {
    result.confidence_and_expectation = payload.confidenceAndExpectation
  }
  if (payload.originalLevel !== undefined) result.original_level = payload.originalLevel || undefined
  if (payload.originalSalary !== undefined) result.original_salary = payload.originalSalary ?? undefined
  if (payload.promotedLevel !== undefined) result.promoted_level = payload.promotedLevel || undefined
  if (payload.promotedBaseSalary !== undefined) {
    result.promoted_base_salary = payload.promotedBaseSalary ?? undefined
  }
  if (payload.promotedPerformanceSalary !== undefined) {
    result.promoted_performance_salary = payload.promotedPerformanceSalary ?? undefined
  }
  if (payload.promotedSalary !== undefined) result.promoted_salary = payload.promotedSalary ?? undefined
  if (payload.selectedApproverUserIds !== undefined) {
    result.selected_approver_user_ids = payload.selectedApproverUserIds || {}
  }
  return result
}

const mapApplicationRecord = (record: ApiPromotionApplicationRecord): PromotionApplicationRecord => ({
  id: record.id,
  applicationNo: record.application_no,
  fillDate: record.fill_date,
  campus: record.campus,
  name: record.name,
  nativePlace: record.native_place,
  age: record.age,
  entryDate: record.entry_date,
  department: record.department,
  position: record.position,
  workOverview: record.work_overview,
  promotionReason: record.promotion_reason,
  confidenceAndExpectation: record.confidence_and_expectation,
  originalLevel: record.original_level,
  originalSalary: record.original_salary,
  promotedLevel: record.promoted_level,
  promotedBaseSalary: record.promoted_base_salary,
  promotedPerformanceSalary: record.promoted_performance_salary,
  promotedSalary: record.promoted_salary,
  createdByUserId: record.created_by_user_id,
  createdByName: record.created_by_name,
  departmentManagerOpinion: record.department_manager_opinion,
  departmentManagerPassed: record.department_manager_passed,
  principalOpinion: record.principal_opinion,
  principalPassed: record.principal_passed,
  bizDirectorOpinion: record.biz_director_opinion,
  bizDirectorPassed: record.biz_director_passed,
  hrDirectorOpinion: record.hr_director_opinion,
  hrDirectorPassed: record.hr_director_passed,
  chairmanOpinion: record.chairman_opinion,
  chairmanPassed: record.chairman_passed,
  isPassed: record.is_passed,
  status: record.status,
  statusLabel: record.status_label,
  currentStage: record.current_stage,
  currentStageLabel: record.current_stage_label,
  rejectionReason: record.rejection_reason,
  submittedAt: record.submitted_at,
  completedAt: record.completed_at,
  canEdit: record.can_edit,
  canDelete: record.can_delete,
  canSubmit: record.can_submit,
  canApprove: record.can_approve,
  selectedApproverUserIds: record.selected_approver_user_ids || {},
  currentApprovers: record.current_approvers.map((item) => ({
    userId: item.user_id,
    name: item.name,
    department: item.department,
    position: item.position,
    campus: item.campus,
  })),
  approvalFlow: record.approval_flow.map((item) => ({
    stage: item.stage,
    stageLabel: item.stage_label,
    status: item.status,
    statusLabel: item.status_label,
    approvers: item.approvers.map((approver) => ({
      userId: approver.user_id,
      name: approver.name,
      department: approver.department,
      position: approver.position,
      campus: approver.campus,
    })),
    action: item.action,
    actionLabel: item.action_label,
    actedByUserId: item.acted_by_user_id,
    actedByName: item.acted_by_name,
    comment: item.comment,
    actedAt: item.acted_at,
  })),
  approvalActions: record.approval_actions.map((item) => ({
    id: item.id,
    stage: item.stage,
    stageLabel: item.stage_label,
    action: item.action,
    approverUserId: item.approver_user_id,
    approverName: item.approver_name,
    comment: item.comment,
    createdAt: item.created_at,
  })),
  createdAt: record.created_at,
  updatedAt: record.updated_at,
})

const mapPreviewStage = (
  stage: ApiPromotionApprovalPreviewStage,
): PromotionApprovalPreviewStage => ({
  stage: stage.stage,
  stageLabel: stage.stage_label,
  recommendedUserIds: stage.recommended_user_ids || [],
  approvers: (stage.approvers || []).map((item) => ({
    userId: item.user_id,
    name: item.name,
    department: item.department,
    position: item.position,
    campus: item.campus,
  })),
})

const mapApprovalConfigRecord = (record: ApiPromotionApprovalConfig): PromotionApprovalConfig => ({
  id: record.id,
  campus: record.campus,
  applyDepartment: record.apply_department,
  applyPosition: record.apply_position,
  stage: record.stage,
  stageLabel: record.stage_label,
  isActive: record.is_active,
  approvers: record.approvers.map((item) => ({
    id: item.id,
    approverUserId: item.approver_user_id,
    approverName: item.approver_name,
    approverDepartment: item.approver_department,
    approverPosition: item.approver_position,
    approverCampus: item.approver_campus,
    sortOrder: item.sort_order,
  })),
  createdAt: record.created_at,
  updatedAt: record.updated_at,
})

export async function listPromotionApplications(params?: {
  campus?: string
  status?: PromotionStatus
  department?: string
}) {
  const response = await api.get<ApiPromotionApplicationRecord[]>('/human-resources/promotion-applications', {
    params: {
      campus: params?.campus || undefined,
      status: params?.status || undefined,
      department: params?.department || undefined,
    },
  })
  return response.data.map(mapApplicationRecord)
}

export async function getPromotionApplication(applicationId: number) {
  const response = await api.get<ApiPromotionApplicationRecord>(
    `/human-resources/promotion-applications/${applicationId}`,
  )
  return mapApplicationRecord(response.data)
}

export async function createPromotionApplication(payload: PromotionApplicationPayload) {
  const response = await api.post<ApiPromotionApplicationRecord>(
    '/human-resources/promotion-applications',
    mapApplicationPayload(payload),
  )
  return mapApplicationRecord(response.data)
}

export async function updatePromotionApplication(
  applicationId: number,
  payload: Partial<PromotionApplicationPayload>,
) {
  const response = await api.put<ApiPromotionApplicationRecord>(
    `/human-resources/promotion-applications/${applicationId}`,
    mapPartialApplicationPayload(payload),
  )
  return mapApplicationRecord(response.data)
}

export async function deletePromotionApplication(applicationId: number) {
  await api.delete(`/human-resources/promotion-applications/${applicationId}`)
}

export async function submitPromotionApplication(applicationId: number) {
  const response = await api.post<ApiPromotionApplicationRecord>(
    `/human-resources/promotion-applications/${applicationId}/submit`,
  )
  return mapApplicationRecord(response.data)
}

export async function approvePromotionApplication(
  applicationId: number,
  payload?: PromotionApprovalActionPayload,
) {
  const response = await api.post<ApiPromotionApplicationRecord>(
    `/human-resources/promotion-applications/${applicationId}/approve`,
    {
      comment: payload?.comment || undefined,
      original_level: payload?.originalLevel || undefined,
      original_salary: payload?.originalSalary ?? undefined,
      promoted_level: payload?.promotedLevel || undefined,
      promoted_base_salary: payload?.promotedBaseSalary ?? undefined,
      promoted_performance_salary: payload?.promotedPerformanceSalary ?? undefined,
      promoted_salary: payload?.promotedSalary ?? undefined,
    },
  )
  return mapApplicationRecord(response.data)
}

export async function rejectPromotionApplication(applicationId: number, comment: string) {
  const response = await api.post<ApiPromotionApplicationRecord>(
    `/human-resources/promotion-applications/${applicationId}/reject`,
    {
      comment,
    },
  )
  return mapApplicationRecord(response.data)
}

export async function listPromotionApprovalConfigs(params?: {
  campus?: string
  applyDepartment?: string
  applyPosition?: string
  stage?: PromotionStage
}) {
  const response = await api.get<ApiPromotionApprovalConfig[]>('/human-resources/promotion-approval-configs', {
    params: {
      campus: params?.campus || undefined,
      apply_department: params?.applyDepartment || undefined,
      apply_position: params?.applyPosition || undefined,
      stage: params?.stage || undefined,
    },
  })
  return (response.data || []).map(mapApprovalConfigRecord)
}

export async function upsertPromotionApprovalConfig(payload: PromotionApprovalConfigPayload) {
  const response = await api.put<ApiPromotionApprovalConfig>('/human-resources/promotion-approval-configs', {
    campus: payload.campus,
    apply_department: payload.applyDepartment || undefined,
    apply_position: payload.applyPosition || undefined,
    stage: payload.stage,
    approver_user_ids: payload.approverUserIds,
    is_active: payload.isActive,
  })
  return mapApprovalConfigRecord(response.data)
}

export async function deletePromotionApprovalConfig(configId: number) {
  const response = await api.delete<{ success: boolean }>(
    `/human-resources/promotion-approval-configs/${configId}`,
  )
  return response.data
}

export async function previewPromotionApproverCandidates(payload: {
  campus: string
  department: string
  position: string
}) {
  const response = await api.post<ApiPromotionApprovalPreviewStage[]>(
    '/human-resources/promotion-applications/approver-preview',
    payload,
  )
  return (response.data || []).map(mapPreviewStage)
}
